### PR TITLE
ID-818, ID-819, ID-916 New Registration Flow, ToS Improvements

### PR DESCRIPTION
**Pull request by Ghost-in-a-Jar**
### Jira Ticket: https://broadworkbench.atlassian.net/browse/ID-916

This is the last step of https://broadworkbench.atlassian.net/browse/ID-787. So, it relies on basically all other tickets being done.

## Summary of changes:
- Users now accept the Terra Terms of Service during registration, instead of as a second step afterwards.
- Terra-UI now supports a Rolling Acceptance Window for Terms of Service update.
- Numerous improvements to user registration and login.

### What
- `Register.tsx` has been updated to prevent a new user from registering until the Terms of Service has been accepted. This is done by disabling the `Register` button until the user has loaded the Terms of Service and checked a checkbox that says the user accepts the Terms of Service. The checkbox is not active until the Terms of Service has been loaded, preventing users from checking the checkbox before they've actually seen the Terms of Service.
- Rendering of remote Markdown content has been encapsulated in `RemoteMarkdown.tsx` 
- `TermsOfServicePage.tsx` has been converted to tsx. The page has been significantly simplified and Grace Period functionality has been removed. If no action is needed, a `Return to Terra` button now appears. The page also uses `RemoteMarkdown` instead of handling the Terms of Service text downloading itself.
- `PrivacyPolicy.tsx` now uses `RemoteMarkdown` instead of handling the Privacy Policy text download itself.
- Old Terms of Service API calls have been removed and replaced with `Ajax().TermsOfService`, which call new Sam endpoints.
- `registrationStatus` has been removed and its functionality incorporated into `signInStatus`. There is now one place to check if the user has authenticated via B2C, registered for Terra, or signed out.
- Rolling Acceptance Window has been implemented as a replacement for Grace Period. The Rolling Acceptance Window allows the user to keep using Terra if the Terms of Service is updated during their active session, but forces them to accept the Terms of Service when they log in. This will ensure that all users are operating under the most recent Terms of Service after 24h of ToS release.

### Why
- Terms of Service must be accepted before we record any information about the user. GDPR compliance! 
- The old Terms of Service update mechanism was clunky, confusing to users, and kicked everyone out of Terra once the Terms of Service update was forced. The new Rolling Acceptance Window is a much better user experience!
- New Sam endpoints that provide for a better user experience are now available.

### Testing strategy
- [x] Unit Tests for new code and functionality.
- [x] `register-user` CUJ test has been updated to test new Terms of Service functionality.
- [x] Extensive manual testing in BEEs and Dev will be done to verify functionality.

 

https://github.com/DataBiosphere/terra-ui/assets/3210510/824b09fc-22c8-451f-9063-8c69b6d11ce7

image
image

### CODE REVIEW

File: src/libs/ajax/TermsOfService.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+export const TermsOfService = (signal?: AbortSignal) => {
+  return {
+    getTermsOfServiceText: async (): Promise<string> => {
+      const res = await fetchSam('termsOfService/v1/docs?doc=termsOfService', { signal });
+      return res.text();
+    },
+    getPrivacyText: async (): Promise<string> => {
+      const res = await fetchSam('termsOfService/v1/docs?doc=privacyPolicy', { signal });
```
Comment (sjkobori) [inline src/libs/ajax/TermsOfService.ts]: Lets make this `getPrivacyPolicyText` to be more consistent.

File: src/libs/ajax/TermsOfService.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      return res.text();
+    },
+    acceptTermsOfService: async (): Promise<void> => {
+      return await fetchSam(
+        'api/termsOfService/v1/user/self/accept',
+        _.mergeAll([authOpts(), { signal, method: 'PUT' }])
+      );
+    },
+    rejectTermsOfService: async (): Promise<void> => {
```
Comment (sjkobori) [inline src/libs/ajax/TermsOfService.ts:33]: Do we want to make these just `accept` and `reject` since they are already namespaced under `TermsOfService`?

File: integration-tests/tests/register-user.js [hunk 3]
```diff
@@ -23,7 +27,15 @@ const testRegisterUserFn = withUser(async ({ page, testUrl, token }) => {
   // We wait here for the cookie message to disappear, it has a fixed duration fading animation of 300ms
   await delay(350);
   await verifyAccessibility(page);
-  await click(page, clickable({ textContains: 'View Workspaces' }));
+
+  // This is the hamburger menu
+  await click(page, '/html/body/div[1]/div[2]/div/div[1]/div[1]/div[1]/div/div[1]');
```
Comment (cahrens) [inline integration-tests/tests/register-user.js]: There has to be a better way!! I would recommend either (in order of preference):

1) Adding an aria-label to the button and using that to select it.
2) Adding a unique ID to the button and using that.
3) Using the button's aria-role to get it (this one could be hard because there are likely multiple such items on the page).

File: integration-tests/tests/register-user.js [hunk 4]
```diff
@@ -34,8 +46,12 @@ const testRegisterUserFn = withUser(async ({ page, testUrl, token }) => {
 
   // Accept the Terms of Service
   await click(page, clickable({ textContains: 'Read Terra Platform Terms of Service here' }));
-  await click(page, clickable({ textContains: 'OK' }));
-  await click(page, clickable({ textContains: 'By checking this box, you are agreeing to the Terra Terms of Service' }));
+  await waitForModal(page);
+  // We seem to need a small delay here
```
Comment (cahrens) [inline integration-tests/tests/register-user.js]: What happens without the delay?

Comment (tlangs) [reply integration-tests/tests/register-user.js]: The test fails to find the OK button and fails.

Comment (tlangs) [reply integration-tests/tests/register-user.js]: ah, ok. `waitForNoSpinners` works here too.

File: integration-tests/tests/register-user.js [hunk 5]
```diff
@@ -23,21 +28,36 @@ const testRegisterUserFn = withUser(async ({ page, testUrl, token }) => {
   // We wait here for the cookie message to disappear, it has a fixed duration fading animation of 300ms
   await delay(350);
   await verifyAccessibility(page);
-  await click(page, clickable({ textContains: 'View Workspaces' }));
+
+  // This is the hamburger menu
+  await click(page, ariaLabel({ labelContains: 'Toggle main menu' }));
```
Comment (cahrens) [inline integration-tests/tests/register-user.js]: I would recommend moving the label to the button so that you can use normal clickable methods. I verified that it still works just fine with the screen reader, and axe reports no issues.

image

Comment (tlangs) [reply integration-tests/tests/register-user.js]: Just tried it, and register-user passes!

File: src/pages/workspaces/workspace/Dashboard/WorkspaceNotifications.test.ts [hunk 6]
```diff
@@ -25,7 +26,15 @@ describe('WorkspaceNotifications', () => {
   };
 
   afterEach(() => {
-    authStore.reset();
+    asMockedFn(Ajax).mockImplementation(
```
Comment (cahrens) [inline src/pages/workspaces/workspace/Dashboard/WorkspaceNotifications.test.ts:29]: If you move this to `beforeEach`, can you then remove the copy down below at line 29?

File: integration-tests/tests/register-user.js [hunk 7]
```diff
@@ -23,21 +27,36 @@ const testRegisterUserFn = withUser(async ({ page, testUrl, token }) => {
...
   await fillIn(page, input({ labelContains: 'Organization' }), 'Test Organization');
   await fillIn(page, input({ labelContains: 'Department' }), 'Test Department');
   await fillIn(page, input({ labelContains: 'Title' }), 'Test Title');
   await fillInReplace(page, input({ labelContains: 'Contact Email for Notifications' }), 'ltcommanderdata@neighborhood.horse');
+
+  // Accept the Terms of Service
+  await click(page, clickable({ textContains: 'Read Terra Platform Terms of Service here' }));
+  await waitForModal(page);
+  await waitForNoSpinners(page);
+  await click(page, button({ text: 'OK' }));
```
Comment (cahrens) [inline integration-tests/tests/register-user.js]: Why did you need to create  a new method for clicking OK? It seems unusual because a lot of the tests click buttons.

Comment (tlangs) [reply integration-tests/tests/register-user.js]: Stemmed from my unfamiliarity with the integration test utils. Removed and replaced with `clickable`.

File: src/pages/workspaces/workspace/Dashboard/WorkspaceNotifications.test.ts (hunk 6 above)
Comment (tlangs) [reply src/pages/workspaces/workspace/Dashboard/WorkspaceNotifications.test.ts:29]: It worked! Done!

**Comment by cahrens** [comment]
Integration test failure in `registerUser`.

File: src/components/TopBar.test.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+const itShowsTheSignInButton = (newState: Partial<AuthState>) => {
+  it('displays the Sign In button in the menu', () => {
+    // Arrange
+    authStore.update((state) => ({ ...state, ...newState }));
+
+    // Act
+    render(h(TopBar));
+
+    // Assert
+    expect(screen.queryByText('Sign In')).not.toBeInTheDocument();
```
Comment (cahrens) [inline src/components/TopBar.test.ts]: This case appears to be the same as the last one… it is verifying that `Sign In` is **not** in the document, which isn't supposed to be true.

File: src/components/TopBar.test.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+type NavExports = typeof import('src/libs/nav');
+jest.mock(
+  'src/libs/nav',
+  (): NavExports => ({
+    ...jest.requireActual<NavExports>('src/libs/nav'),
+    getLink: jest.fn(),
+  })
+);
+
+const itDoesNotShowTheSignInButton = (newState: Partial<AuthState>) => {
```
Comment (cahrens) [inline src/components/TopBar.test.ts]: This is a somewhat unusual way to write these tests. Take a look at `it.each`, which is a more standard way to iterate though variations.

Example: https://github.com/DataBiosphere/terra-ui/blob/a9fe75a4cac8424376addf4c1a3b34ca42172c44/src/libs/utils.test.ts#L41

**Comment by cahrens** [comment]
The lack of margins and lack of a landmark on the privacy page is not great.

image

**Comment by tlangs** [comment]
The privacy page has always been like that, unfortunately. We could make some improvements in this PR

**Review by cahrens — APPROVED** [review]
Approving the minimal workspace and branding file changes, but please do fix the unittest that isn't testing what it is intended to, https://github.com/DataBiosphere/terra-ui/pull/4482#discussion_r1402543177

I skimmed some portions of the other code, but am relying on Identiteam reviewers to review all the files in depth.

**Comment by cahrens** [comment]
Integration test failure in `registerUser` appears to be the same as the last one. Note that you should be able to use the [test flakes](https://github.com/DataBiosphere/terra-ui/wiki/Integration-tests#detecting-flakiness) to run the integration test a bunch of times locally to ensure that it passes consistently (like 50/50 would be good).

**Comment by tlangs** [comment]
Increased a timeout on `register-user` CUJ, then ran `yarn test-flakes register-user`. Out of 100 runs, 100 passed!

**Review by tlangs — APPROVED** [review]

**Review by Shakespeared — APPROVED** [review]
